dimod/utilities.py

```diff
@@ -340,3 +345,48 @@ def resolve_label_conflict(mapping, old_labels=None, new_labels=None):
             # don't need to add it to intermediate_to_new because it is a self-label
 
     return old_to_intermediate, intermediate_to_new
+
+
+def child_structure_dfs(sampler, seen=None):
+    """Find the structure of a composed sampler by a depth first search on the
+    children.
+
+    Args:
+        sampler (:obj:`.Sampler`):
+            A :class:`.Structured` sampler or a composed sampler with at least
+            one structured child.
+
+        seen (set, optional, default=False):
+            A set of the ids of the child samplers that have already been
+            checked.
+
+    Returns:
+        :class:`~collections.namedtuple`: A named tuple of the form
+        `Structure(nodelist, edgelist, adjacency)`, where the 3-tuple values
+        are the :attr:`.Structured.nodelist`, :attr:`.Structured.edgelist`
+        and :attr:`.Structured.adjacency` attributes of the first structured
+        sampler encountered.
+
+    """
+    seen = set() if seen is None else seen
```

Comment:
Why is `seen` needed for DFS? Do we want to support chains of composites with cyclic references (`a = A(B(a))`)? Those seem infinitely-evaluating, thus invalid by default.

Reply:
It would be possible to construct a set of samplers/composites for which you can get cycles, though I agree all of the cases I can think of for which that's true are arbitrary and would not be useful. But since it's possible, and keeping the set of samplers available is cheap, why not?

Reply:
Well, it makes the code more complicated and it's uglier. :smile: 
But that's fine if we actually want to handle cyclic refs.